### DISCUSSION TITLE
Is there a way to improve citnames build speed

### DISCUSSION THREAD

**Discussion by SunJun8** [rizsotto/Bear#451] [Q&A]
As the title,

it takes a long time to build json for a 300M compiled database, and citnames seems to use only one cpu core

**Comment by rizsotto** [comment]
Sure, that process is a single threaded one. And currently does not do anything which would block parallel processing.

Are you interested to implement it? (I can guide you what's needed to make it.)

**Reply by SunJun8** [reply]
> Sure, that process is a single threaded one. And currently does not do anything which would block parallel processing.
> 
> Are you interested to implement it? (I can guide you what's needed to make it.)

ok i would like to try

**Reply by rizsotto** [reply]
Thanks @SunJun8 !

The function which needs a change is the [`transform`](https://github.com/rizsotto/Bear/blob/master/source/citnames/source/Application.cc#L164) function in the `source/citnames/source/Application.cc`. As you can see, it does transform each event into one (or more) compilation database entries, and merge the together.

I've tried to use [`std::transform_reduce`](https://en.cppreference.com/w/cpp/algorithm/transform_reduce) to run it parallel, but looks like parallel-STL functions are still not ready. So, I would recommend to write a similar map-reduce like higher-order function, which is written by the existing C++17 constructs, and can be use to implement the `transform` method.

**Reply by rizsotto** [reply]
Looking for help to make this work. https://github.com/rizsotto/Bear/pull/455 is a PR where I use parallel STL, but see no improvements. If anyone knows how to fix it, I would be glad.

**Comment by rizsotto** [comment]
This might be fixed by `4.0.0` version. The event processing is done during the interception.
